### DISCUSSION TITLE
Variadic static constructors ?

### DISCUSSION THREAD

**Discussion by drupol** [loophp/collection#129] [Ideas]
I'm just throwing this here so I don't forget!

What do you think about having variadic static constructors?

from:

`    public static function fromIterable(iterable $iterable): self`

to:

`    public static function fromIterable(iterable ...$iterable): self`

**Comment by AlexandruGG** [comment]
interesting idea. Is there a specific use case where you thought it would be helpful?

My concerns with this are:
- will it work for all the static constructors? I'm thinking it might not work very easily with some of them
- it might confuse users depending on which approach we take with this. For example:

```PHP
$col1 = Collection::fromIterable([1, 2, 3]); // this is clear, I get Collection([1, 2, 3])

/**
* This one is not super clear, am I getting:
* A) Collection([1, 2, 3, 4, 5, 6]) -> the iterables get merged
* B) Collection(Collection([1, 2, 3]), Collection([4, 5, 6])) -> I get a collection where each iterable gets put into a collection
* C) Collection([[1, 2, 3], [4, 5, 6]]) -> I get a collection where each iterable is kept in a separate iterable (e.g. an array)
*/ 
$col2 = Collection::fromIterable([1, 2, 3], [4, 5, 6]);

**Reply by drupol** [reply]
Basically, I'm thinking about this... it would be based on `MultipleIterableIterator` instead of `IterableIterator`.

```php
public static function fromIterable(iterable ...$iterables): self
{
    return new self(
        static fn (iterable ...$iterables): Iterator => new MultipleIterableIterator(...$iterables),
        ...$iterables
    );
}
```

**Reply by AlexandruGG** [reply]
I see, I suspected this. So effectively it will behave like scenario `A)` ?

**Reply by drupol** [reply]
Yep

**Comment by AlexandruGG** [comment]
So on this, would it be for other constructors or only for `fromIterable`? Could you make a list with the ones that would be doable?

**Reply by drupol** [reply]
I think it could be done for:

* `fromFile`
* `fromIterable`
* `fromResource`
* `fromString`

**Reply by AlexandruGG** [reply]
for `fromString` the second argument is the **delimiter**. Would this have to move to being the first argument and the variadic strings to the second?

**Reply by drupol** [reply]
Oops, indeed! Then we have to remove `fromString` of that list.
